src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+                    return ConsumeNumberResult.NeedMoreData;
+            }
+            if (JsonConstants.Delimiters.IndexOf(nextByte) >= 0)
+                return ConsumeNumberResult.Success;
+
+            return ConsumeNumberResult.OperationIncomplete;
+        }
+
+        private ConsumeNumberResult ConsumeDecimalDigits(ref ReadOnlySpan<byte> data, ref int i, byte nextByte)
+        {
+            if (i >= data.Length)
+            {
+                if (IsLastSpan)
+                {
+                    _lineBytePosition += i;
+                    ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedDigitNotFoundEndOfData, nextByte);
+                }
+                else
+                    return ConsumeNumberResult.NeedMoreData;
+            }
+            nextByte = data[i];
+            if ((uint)(nextByte - '0') > '9' - '0')
+            {
+                _lineBytePosition += i;
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedDigitNotFound, nextByte);
+            }
+            i++;
+
+            return ConsumeIntegerDigits(ref data, ref i);
+        }
+
+        private ConsumeNumberResult ConsumeSign(ref ReadOnlySpan<byte> data, ref int i, byte nextByte)
+        {
+            if (i >= data.Length)
+            {
+                if (IsLastSpan)
+                {
+                    _lineBytePosition += i;
+                    ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedDigitNotFoundEndOfData, nextByte);
+                }
+                else
+                    return ConsumeNumberResult.NeedMoreData;
+            }
+
+            nextByte = data[i];
+            if (nextByte == '+' || nextByte == '-')
+            {
+                i++;
+                if (i >= data.Length)
+                {
+                    if (IsLastSpan)
+                    {
+                        _lineBytePosition += i;
+                        ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedDigitNotFoundEndOfData, nextByte);
+                    }
+                    else
+                        return ConsumeNumberResult.NeedMoreData;
+                }
+                nextByte = data[i];
+            }
+
+            if ((uint)(nextByte - '0') > '9' - '0')
+            {
+                _lineBytePosition += i;
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedDigitNotFound, nextByte);
+            }
+
+            return ConsumeNumberResult.OperationIncomplete;
+        }
+
+        /// <summary>
+        /// This method consumes the next token regardless of whether we are inside an object or an array.
+        /// For an object, it reads the next property name token. For an array, it just reads the next value.
+        /// </summary>
+        private ConsumeTokenResult ConsumeNextToken(byte marker)
+        {
+            // TODO: Avoid unconstrained recursive calls
```

Comment:
Well, you clearly already know it, but: "Avoid unconstrained recursive calls" :smile: